tools/codegen/model.py

```diff
@@ -47,11 +47,89 @@ def __str__(self) -> str:
 # Valid values of the 'variants' field in native_functions.yaml
 Variant = Enum('Variant', ('function', 'method'))
 
+# NOTE: Keep the list in sync with `DispatchKey` in c10/core/DispatchKey.h
```

Comment:
I wonder if it's possible to add a check to enforce the parity between the two lists? Otherwise the registration can easily go to the wrong backends. :P

Reply:
@ailzhang I think we're actually safe from that, since we're not generating the enum's underlying values into code - it's still the symbolic names via `__str__` (hence byte-for-byte compatible output).

That said, we could think about adding parity enforcement later on - say, a static assert that verified the same total number of distinct dispatch keys - not sure the best generated code to add this to though.

Better yet would be a single source of truth. I think our two alternatives are
* actually codegen the C++ definition (I think we have a cutout that prevents regeneration of unchanged artifacts that we could use, foggy on details though)
* read the C++ definition into python, though doing this without actually parsing C++ would probably involve changes that would degrade UX on the C++ side (macros or whatever)

Anyway though, given that this PR makes things strictly better than before (AFAICT codegen used to just silently ignore invalid dispatch key names in `dispatch:` entries), I don't think any of the above is worth piling onto this PR unless it's dead easy/obvious.

Reply:
> actually codegen the C++ definition 

...this fits in with the broader idea of lifting the entire dispatch key data model into metadata and using it to generate not just this enum but also helper functions and so on

Reply:
Thanks @bhosmer for the rely - I was about to comment the same :p

I was thinking of adding higher level parity check in subsequent PRs where I do dispatch table computation.
I also thought about codegen / macro + parsing / macro + test-method-to-print-all. Might do one of these later :)